blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala

```diff
@@ -80,18 +74,20 @@ class Http4sWSStage[F[_]](ws: ws4s.Websocket[F])(implicit F: Effect[F], val ec:
     val onStreamFinalize: F[Unit] =
       for {
         dec <- F.delay(count.decrementAndGet())
-        _ <- deadSignal.map(signal => if (dec == 0) signal.set(true))
+        _ <- if (dec == 0) deadSignal.set(true) else ().pure[F]
       } yield ()
 
     // Effect to send a close to the other endpoint
     val sendClose: F[Unit] = F.delay(sendOutboundCommand(Command.Disconnect))
 
-    val wsStream = for {
-      dead <- deadSignal
-      in = inputstream.to(ws.receive).onFinalize(onStreamFinalize)
-      out = ws.send.onFinalize(onStreamFinalize).to(snk).drain
-      merged <- in.mergeHaltR(out).interruptWhen(dead).onFinalize(sendClose).compile.drain
-    } yield merged
+    val wsStream = inputstream
+      .to(ws.receive)
+      .onFinalize(onStreamFinalize)
+      .mergeHaltR(ws.send.onFinalize(onStreamFinalize).to(snk).drain)
```

Comment:
`mergeHaltR(foo.drain)` can probably be `concurrently(foo)` now

Reply:
Actually scratch that sorry, you nee the output of `foo`, but want to complete when `bar` is done